cdap-common/src/main/java/co/cask/cdap/common/zookeeper/ZKExtOperations.java

```diff
@@ -353,58 +274,293 @@ public void onFailure(Throwable t) {
+    Futures.addCallback(dataSupplier.get(), new FutureCallback<T>() {
+      @Override
+      public void onSuccess(final T data) {
+        try {
+          // Try to create the node without creating parent. This is to make sure the latest data
+          // is being used if there are concurrent modification to the node (CDAP-4388)
+          OperationFuture<String> createFuture = (acls == null)
+            ? zkClient.create(path, codec.encode(data), CreateMode.PERSISTENT, false)
+            : zkClient.create(path, codec.encode(data), CreateMode.PERSISTENT, false, acls);
 
-            OperationFuture<String> future;
-            if (createAcl == null) {
-              future = zkClient.create(path, data, CreateMode.PERSISTENT);
-            } else {
-              future = zkClient.create(path, data, CreateMode.PERSISTENT, createAcl);
+          Futures.addCallback(createFuture, new FutureCallback<String>() {
+            @Override
+            public void onSuccess(String result) {
+              // If creation succeeded, the operation is completed
+              resultFuture.set(data);
             }
 
-            Futures.addCallback(future, new FutureCallback<String>() {
-              @Override
-              public void onSuccess(String result) {
-                resultFuture.set(content);
-              }
+            @Override
+            public void onFailure(Throwable t) {
+              if (t instanceof KeeperException.NoNodeException) {
+                // If failed with NoNode, it means parent path doesn't exist. Create the parent path first and retry
+                OperationFuture<String> createParentFuture = zkClient.create(getParent(path), null,
+                                                                             CreateMode.PERSISTENT);
+                Futures.addCallback(createParentFuture, new FutureCallback<String>() {
+                  @Override
+                  public void onSuccess(String result) {
+                    // If creation of parent path succeeded, try to create again.
+                    // We call the create method again so that the dataSupplier will be called to get the latest value.
+                    create(zkClient, path, dataSupplier, codec, acls, resultFuture);
+                  }
 
-              @Override
-              public void onFailure(Throwable t) {
-                if (t instanceof KeeperException.NodeExistsException) {
-                  // If failed to create due to node exists, try to do getAndSet.
-                  getAndSet(zkClient, path, modifier, codec, resultFuture, createAcl);
-                } else {
-                  resultFuture.setException(t);
-                }
+                  @Override
+                  public void onFailure(Throwable t) {
+                    // If failed to create parent path, fail the operation
+                    resultFuture.setException(t);
```

Comment:
We do. See https://github.com/caskdata/cdap/pull/5741/files/211dc93c21205909f3eae7a5a1c313ea0fa4a26a#diff-eb718f398c9da13e9b39257952275516R461